lib/ukstore/include/uk/store.h

```diff
@@ -74,207 +74,272 @@ enum uk_store_entry_type {
-		.set.e_type = (e_set),					\
-		.flags      = UK_STORE_ENTRY_FLAG_STATIC,		\
-		.cookie     = (arg_cookie)				\
+		.id             = eid,					\
+		.name           = STRINGIFY(entry),			\
+		.type           = (UK_STORE_ENTRY_TYPE(e_type)),	\
+		.ops.get.e_type = (e_get),				\
+		.ops.set.e_type = (e_set),				\
+		.flags          = UK_STORE_ENTRY_FLAG_STATIC		\
 	}
 
 /* Do not call directly */
-#define _UK_STORE_STATIC_ENTRY(entry, lib_str, e_type, e_get, e_set,	\
-				cookie)					\
-	__UK_STORE_STATIC_ENTRY(entry, lib_str, e_type, e_get, e_set, cookie)
+#define _UK_STORE_STATIC_ENTRY(id, entry, lib_str, e_type, e_get, e_set)\
+	__UK_STORE_STATIC_ENTRY(id, entry, lib_str, e_type, e_get, e_set)
 
 /**
  * Adds an entry to the entry section of a library.
  *
+ * @param id entry id
  * @param entry the entry in the section
  * @param e_type type for the entry, e.g., s8, u16, charp
  * @param e_get getter pointer (optional, can be NULL)
  * @param e_set setter pointer (optional, can be NULL)
- * @param cookie a cookie for extra storage
  */
-#define UK_STORE_STATIC_ENTRY(entry, e_type, e_get, e_set, cookie)	\
-	_UK_STORE_STATIC_ENTRY(entry, STRINGIFY(__LIBNAME__),		\
-		e_type, e_get, e_set, cookie)
+#define UK_STORE_STATIC_ENTRY(id, entry, e_type, e_get, e_set)		\
+	_UK_STORE_STATIC_ENTRY(id, entry, STRINGIFY(__LIBNAME__),	\
+			       e_type, e_get, e_set)
 
 #else /* !CONFIG_LIBUKSTORE */
 
 /* Library-specific information generated at compile-time */
 #include <uk/bits/store_libs.h>
 
+struct uk_store_ops {
+	/* Function getter pointer */
+	union {
+		_uk_store_get_s8_func_t     s8;
+		_uk_store_get_u8_func_t     u8;
+		_uk_store_get_s16_func_t    s16;
+		_uk_store_get_u16_func_t    u16;
+		_uk_store_get_s32_func_t    s32;
+		_uk_store_get_u32_func_t    u32;
+		_uk_store_get_s64_func_t    s64;
+		_uk_store_get_u64_func_t    u64;
+		_uk_store_get_uptr_func_t   uptr;
+		_uk_store_get_charp_func_t  charp;
+	} get;
+
+	/* Function setter pointer */
+	union {
+		_uk_store_set_s8_func_t    s8;
+		_uk_store_set_u8_func_t    u8;
+		_uk_store_set_s16_func_t   s16;
+		_uk_store_set_u16_func_t   u16;
+		_uk_store_set_s32_func_t   s32;
+		_uk_store_set_u32_func_t   u32;
+		_uk_store_set_s64_func_t   s64;
+		_uk_store_set_u64_func_t   u64;
+		_uk_store_set_uptr_func_t  uptr;
+		_uk_store_set_charp_func_t charp;
+	} set;
+
+	_uk_store_entry_freed_func_t cleanup;
+};
+
+struct uk_store_entry {
+	/* Entry unique id */
+	__u64 id;
+
+	/* The entry name */
+	char *name;
+
+	/* Entry flags */
+	int flags;
+
+	/* Entry getter/setter type */
+	enum uk_store_entry_type type;
+
+	/* Entry ops */
+	struct uk_store_ops ops;
+} __align8;
+
+struct uk_store_object {
+	/* List for the dynamic objects only */
+	struct uk_list_head object_head;
+
+	/* List for the entries in the object */
+	struct uk_list_head entry_head;
+
+	/* Allocator to be used in allocations */
+	struct uk_alloc *a;
+
+	/* Object unique id */
+	__u64 id;
+
+	/* The object name */
+	char *name;
```

Comment:
These cannot be `const char *` without changing the implementation because `uk_store_allocate_object()` and `_UK_STORE_DYNAMIC_CREATE_TYPED` copy the `const char *` name passed by the caller into a locally allocated buffer. Not sure why it was done like that, so I assumed there was good reason to do so (although I still don't see it).

Reply:
This makes sense actually. An API user likely generates object names from a template and you do not want to deal with keeping track of the string buffer.